### PR TITLE
export parameter bindings to ssv file

### PR DESCRIPTION
**Pull request by arun3688**
### Related Issues

https://github.com/OpenModelica/OMSimulator/issues/735

### Purpose
This PR will now export parameter bindings to a ssv file by default.  We can still export inline parameter bindings by setting the  config flag "--exportParametersInline=true"

### Approach
Model.cpp   - Added top level ssvdoc, which exports parameter bindings to ssv file,  and also update the ssd file with top level parameter bindings something like this (e.g)  `<ParameterBinding source="resources/ControlledTemperature.ssv">` .

Parameters.cpp  - added new functions   `exportToSSV() and exportStartValuesHelper`

### CODE REVIEW

**Comment by arun3688** [comment]
@lochel importing parameterBindings from ssv file at the top level requires parameter mapping i guess, we need to discuss about this  see https://github.com/OpenModelica/OMSimulator/pull/755/commits/8eb314ba03af126922837bfea8543637117a7ee5

**Comment by robha67** [comment]
If i have understod the SSP standard correctly then the bindnings are considered via the System Structure Mapping (SSM) format. Have you looked into this?

**Comment by arun3688** [comment]
@robha67 Yes that is correct, we have to perform the mapping via ssm file or inline parameterMapping

File: src/OMSimulatorLib/Model.cpp [hunk 1]
```diff
@@ -264,6 +288,36 @@ oms_status_enu_t oms::Model::list(const oms::ComRef& cref, char** contents)
...
+  if (parameterNodeCount > 0 && isTopSystemOrModel)
+  {
+    // update the ssd with the top level parameterBindings (e.g)  <ParameterBinding source="resources/ControlledTemperature.ssv">
+    for(pugi::xml_node_iterator it = node.begin(); it != node.end(); ++it)
+    {
+      if (std::string(it->name()) == oms::ssp::Draft20180219::ssd::connectors) // insert the parameter bindings after top-level connectors node
+      {
+        pugi::xml_node node_parameters_bindings = node.insert_child_after(oms::ssp::Version1_0::ssd::parameter_bindings, *it);
+        pugi::xml_node node_parameter_binding  = node_parameters_bindings.append_child(oms::ssp::Version1_0::ssd::parameter_binding);
+        std::string ssvFileName = std::string(this->getCref()) + ".ssv";
```
Comment (lochel) [inline src/OMSimulatorLib/Model.cpp:311]: Shouldn't it be `resources/<name>.ssv`?

File: src/OMSimulatorLib/Model.cpp [hunk 2]
```diff
@@ -478,6 +572,10 @@ oms_status_enu_t oms::Model::exportToFile(const std::string& filename) const
   //        -9  Compress better
   //        -j  exclude path. store only the file name
   std::vector<std::string> resources;
+  // TODO, export ssv file to resource directory
+  if (parameterNodeCount > 0)
```
Comment (lochel) [inline src/OMSimulatorLib/Model.cpp:576]: Shouldnb't you check for the filename? E.g. `if (!ssvFileName.empty())`

File: src/OMSimulatorLib/Model.cpp (hunk 1 above)
Comment (arun3688) [reply src/OMSimulatorLib/Model.cpp:311]: This is correct , now the ssv file is exported to resource directory

File: src/OMSimulatorLib/Model.cpp (hunk 2 above)
Comment (arun3688) [reply src/OMSimulatorLib/Model.cpp:576]: Yes we can check also for ssvFileName is empty,  but both are doing the same, i change to check for ssvFile

File: src/OMSimulatorLib/ComponentFMUCS.cpp [hunk 3]
```diff
@@ -752,7 +775,11 @@ oms_status_enu_t oms::ComponentFMUCS::setBoolean(const ComRef& cref, bool value)
     return logError_UnknownSignal(getFullCref() + cref);
 
   if (oms_modelState_virgin == getModel()->getModelState())
-    startValues.setBoolean(allVariables[j].getCref(), value);
+    if (Flags::ExportParametersInline())
+      startValues.setBoolean(allVariables[j].getCref(), value);
+    else
+      // append startValues with prefix (e.g) addP.K1
+      startValues.setBoolean(getCref()+allVariables[j].getCref(), value);
```
Comment (lochel) [inline src/OMSimulatorLib/ComponentFMUCS.cpp:782]: I think we should put curly brace around this to make it easier to read:
```C
if (oms_modelState_virgin == getModel()->getModelState())
{
    if (Flags::ExportParametersInline())
      startValues.setBoolean(allVariables[j].getCref(), value);
    else
      startValues.setBoolean(getCref() + allVariables[j].getCref(), value);
}

Comment (arun3688) [reply src/OMSimulatorLib/ComponentFMUCS.cpp]: added the curly brackets to improve readability of conditions statements see https://github.com/OpenModelica/OMSimulator/pull/755/commits/c55abe718394f2adc80c6c6ec71331f4b8629098

**Comment by markaren** [comment]
Did this PR change how SSPs are imported? 
Because I'm able to simulate my SSP using `OMSimulator-win64-v2.1.0-dev-174-gdbf5fae`, but not `OMSimulator-win64-v2.1.0-dev-176-g00f5648` and `OMSimulator-win64-v2.1.0-dev-180-ge53984d`

```
info:    Initialized thread pool for model QuarterTruck with 2 threads
error:   [setReal] Unknown signal "QuarterTruck.QuarterTruckSystem.wheel.dWheel"
error:   [instantiate] Failed to set start value for R.dWheel
error:   [initialize] Model "QuarterTruck" is in wrong model state
error:   [simulate] Model "QuarterTruck" is in wrong model state
error:   [terminate] Model "QuarterTruck" is in wrong model state
error:   [terminate] Model "QuarterTruck" is in wrong model state
./run.sh: line 2:  2020 Segmentation fault      OMSimulator -t=1000 --numProcs=2 ../QuarterTruck_10.ssp
```

**Comment by arun3688** [comment]
@markaren Yes this PR changed the importing of SSP, But i think the issue might be related to this PR https://github.com/OpenModelica/OMSimulator/pull/756, Can you attach the ssp file so that i can have a look into it.

**Pull request merged**